File: trees/btree_height.py
from trees import TreeNode, build_tree

"""
Test Cases
4
1 2 3
2 N 1 3 N
1 2 N 3 N 4 N
1 2 3 N N 4 6 N 5 N N 7 N

Output
2
3
4
5
"""
def get_height(root: TreeNode) -> int:
    if root is None:
        return 0
    return max(get_height(root.left), get_height(root.right)) + 1


def main():
    test_cases: int = int(input())
    results = []
    for _ in range(test_cases):
        tree_str: str = input()
        root: TreeNode = build_tree(tree_str)
        height: int = get_height(root)
        results.append(height)
    for res in results:
        print(res)

if __name__ == "__main__":
    main()

File: rust/src/bin/btree_height.rs
/*
Test Cases
4
1 2 3
2 N 1 3 N
1 2 N 3 N 4 N
1 2 3 N N 4 6 N 5 N N 7 N
*/

use std::io::{self, BufRead};
use tree::{build_tree, TreeLink};

fn get_height(root: &TreeLink) -> usize {
    match root {
        None => 0,
        Some(node) => {
            let left = node.borrow().left.clone();
            let right = node.borrow().right.clone();
            1 + get_height(&left).max(get_height(&right))
        }
    }
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();
    let t: usize = lines.next().unwrap().unwrap().trim().parse().unwrap();
    let mut results = Vec::new();
    for _ in 0..t {
        let line = lines.next().unwrap().unwrap();
        let root = build_tree(line.trim());
        results.push(get_height(&root));
    }
    for h in &results {
        println!("{}", h);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tree::build_tree;

    #[test]
    fn test_height() {
        assert_eq!(get_height(&build_tree("1 2 3")), 2);
        assert_eq!(get_height(&build_tree("2 N 1 3 N")), 3);
        assert_eq!(get_height(&build_tree("1 2 N 3 N 4 N")), 4);
        assert_eq!(get_height(&build_tree("1 2 3 N N 4 6 N 5 N N 7 N")), 5);
    }
}
